frontend_tests/node_tests/composebox_typeahead.js

```diff
@@ -1410,12 +1411,16 @@ test("begins_typeahead", ({override, override_rewire}) => {
     assert_typeahead_equals("<timerandom", false);
     assert_typeahead_equals("<time", ["translated: Mention a time-zone-aware time"]);
     assert_typeahead_equals("<time:", ["translated: Mention a time-zone-aware time"]);
+    assert_typeahead_equals(":octopus: <time:", ["translated: Mention a time-zone-aware time"]);
```

Comment:
Good job with adding tests.

Having one test case here is insufficient.

See lines 1410-1412 in the original. They're testing 3 scenarios for what comes after "<time":
- "random"
- nothing
- ":"

To catch all the issues, you'll also need to cover those cases.
For example, from reading the code change, I can tell that `":octopus: <timerandom"` will fail.

Please try fixing things up along with these test cases:
- `":octopus: <timerandom"`
- `":octopus: <time"`
- `":octopus: <time:"`

Reply:
Good call, I added a few more tests along these lines! During that, I found the test case you suggest will fail does not, but `":octopus: <tim"` consistently returns the emoji picker results. Additionally, if I add `":octopus: "` in the emoji test section, it fails, but it seems to me like it should be completed already and shouldn't create an emoji one.

That seems like it might be a bug, one of the other ones you mentioned. If so, how would you suggest we proceed?

For now I've left commented out tests with the failing test cases, not intending to merge those.

Reply:
sorry, I thought I tested it out and it failed, but I remembered wrong.

Let's wait for someone from Zulip who knows the intricacies of `composebox_typeahead.get_candidates()` and `composebox_typeahead.filter_and_sort_candidates()` to review this.

The `tokenize_compose_str()` doesn't tokenize based on a well defined lexical grammar. It's harder for newcomers to fully understand.

`":octopus: <tim"` confusingly returning emoji list is because `tokenize_compose_str()` doesn't handle the space character at all. Even though a comment says maybe it should.
https://github.com/zulip/zulip/blob/932ce6cee490d01c376e0ab66944ab0f981bf5e5/static/js/composebox_typeahead.js#L279-L281

And then later on in `composebox_typeahead.get_candidates()`, the code just checks to see if the token starts with ":" and isn't followed by a space.